libyul/backends/wasm/polyfill/Interface.yul

```diff
@@ -51,7 +51,8 @@ function caller() -> z1, z2, z3, z4 {
 
 function callvalue() -> z1, z2, z3, z4 {
 	eth.getCallValue(0:i32)
-	z1, z2, z3, z4 := mload_internal(0:i32)
+	z3 := i64.load(8:i32)
```

Comment:
This does not need a bswap?

Reply:
Yeah, `mload_internal()` does the endianness swap internally. Why doesn't it need it? And why for example `balance()` still does?

Reply:
I just checked the hera implementation `EthereumInterface::eeiGetCallValue(uint32_t resultOffset)` is just using `storeUint128(m_msg.value, resultOffset);` to write the actual value. `storeUint128` uses `storeMemoryReverse` to store the actual value. That means after storing this value seem to be big-endian. Because of the usage of `storeUint128`, the actual value is big-endian in memory, therefore the usage of bswap should be needed here. The strange thing is that the tests only run correctly, if we don't do the byte-swap.

So I thought that we store the value wrong within the call polyfill. After checking I saw we store it correctly, we retrieve the value and use `mstore_internal`. That means we definitely should have the value as big-endian in memory:
```
	let v1, v2 := u256_to_u128(c1, c2, c3, c4)
	mstore_internal(32:i32, 0, 0, v1, v2)
	// ...
	x4 := i64.extend_i32_u(eth.call(g, 12:i32, 32:i32, to_internal_i32ptr(d1, d2, d3, d4), u256_to_i32(e1, e2, e3, e4)))
```

So I checked then what hera is actually doing in `eeiCall`. Here we load the memory correctly `call_message.value = loadUint128(valueOffset);`, during the load we also swap big-endian back to little-endian, because `loadUint128` uses `loadMemoryReverse`.

However, I noticed that `eeiCall` gets never executed. So I took a look at the `ExecutionFramework`, especially `ExecutionFramework::sendMessage`. Here we setup the message value with  `message.value = EVMHost::convertToEVMC(_value);`. And looking at `message.value`, it is already big-endian (the higher addresses store smallest parts of the value). That means if this value is stored using `storeUint128` the bytes got swapped again to little-endian. Thats why I noticed that the `bswap` is not needed here.

I guess that the message values need to be in big-endian, before they arrive at an vm. If we would change it here, the current tests via evm-one would also fail. I think the interaction with an evm from the test-framework perspective should be identical, for the evmone- and the ewasm-case. So I think it is a problem in hera.

I will take a look what is the best option to fix hera tomorrow.

@axic would you agree that this is a hera problem?

Reply:
From https://github.com/ewasm/design/blob/master/eth_interface.md:
> u128: a 128 bit number, represented as a 16 bytes long little endian unsigned integer in memory
>
> getExternalBalance
> Gets balance of the given account and loads it into memory at the given offset.
>
> Parameters
>
> addressOffset i32ptr the memory offset to load the address from (address)
> resultOffset i32ptr the memory offset to load the balance into (u128)

The byteswapping in `storeUint128` is because EVMC considers the value big endian, so this swaps it for wasm.